### ISSUE TITLE
Sanction Module

### ISSUE THREAD

**Issue by SpicyLemon** [provenance-io/provenance#1306]
## Summary

Create a sanction module that allows, through governance proposals, sanctioning of accounts. Funds in sanctioned accounts are frozen and cannot be sent, spent, moved, used, etc.

## Problem Definition

As a regulated entity trying to do business using a blockchain, I want to be able to lock an account, preventing funds from being sent from it, so that the chain can comply with government sanctions and I can continue to use the blockchain within the law.

## Proposal

The `x/sanction` module will house the stuff needed to manage a list of sanctioned accounts. Sanctioning accounts will be done through a governance proposals.

### Concepts

Sanctioning an account often needs to happen faster than a governance proposal vote period though. As such, this module provides the ability to make sanctions happen immediately upon proposal, but require a significantly larger deposit. If the proposal passes, the sanction stands and the deposit is returned. If the proposal does not pass, the effects are undone and the deposit is burned.

The module will also contain a list of unsanctionable addresses. This is to prevent, e.g. the primary fee pool from being sanctioned. This list is provided to the module/keeper when it is created and is not manageable through any interactions, e.g. Txs.

When an address is sanctioned, no funds can be removed from the account.

The ability to immediately unsanction an address exists to counteract a bad actor that issues an immediate sanction against a critical account, even though they expect the vote will fail and they will lose their deposit. However, the deposit required for it should be higher even than an immediate sanction in order to prevent a sanctioned bad actor from having their account unsanctioned, even if only temporarily while it’s being voted on.

The minimum deposit needed for an immediate sanction or immediate unsanction is maintained as module params as two separate values.

When a proposal is submitted, if it has enough total deposit to be immediate, the immediate temporary sanctions or unsanctions are enacted. Similarly, if extra deposit is added to a previously submitted proposal that gives it enough total deposit, the immediate entries are created. Once a proposal is done, these immediate temporary entries are removed regardless of proposal outcome.

Checking of the deposit and enacting immediate temporary sanctions/unsanctions is handled by hooks on the governance module. Cleanup of temporary entries is also handled by governance hooks.

### Governance Proposals

1. Sanction Addresses
    Contents: A list of addresses, an authority.
2. Unsanction Addrsses
    Contents: A list of addresses, an authority.
3. Update Params
    Contents: Params, an authority.

### Query Endpoints

1. Is Sanctioned
    Input: An address.
    Output: True if sanctioned, false if not.
2. Sanctioned Addresses
    Input: Pagination params.
    Output: A list of addresses and pagination info.
3. Temporary Entries
    Input: Pagination params.
    Output: A list of temporary entries and pagination info. A temporary entry has an address, prop id and status (sanction/unsanction).
4. Params
    Input: Nothing.
    Output: Module params (either the ones stored in state or otherwise the defaults).

### Updates to `x/bank`

1. Add a sanction keeper to the bank's send keeper. That "sanction keeper" should be an expected interface with an `IsSanctionedAddr(ctx, addr)` function. I.e. the bank module should not have any actual dependencies (imports) on the sanction module.
2. Add a `SetSanctionKeeper` function to the bank's send keeper. Calling it should update the existing keeper (i.e. don't make it like `WithMintCoinsRestriction`).
3. In `subUnlockedCoins`, check if the address is sanctioned and return an error if it is. This checking should be skipped if the sanction keeper has not been provided.

____

#### For Admin Use

- [ ] Not duplicate issue
- [ ] Appropriate labels applied
- [ ] Appropriate contributors tagged
- [ ] Contributor assigned/self-assigned

**Comment by SpicyLemon** [comment]
This issue was split out from #1046 which covers both sanction and quarantine.

This hackmd was also referenced: https://hackmd.io/@dwedul/BJuR4aQDj
The contents in here are slightly different though due to discoveries and changes made during implementation: https://github.com/provenance-io/cosmos-sdk/pull/401

**Comment by SpicyLemon** [comment]
While working on this, I thought about creating a new account type (similar to vesting accounts) that a sanctioned account would be changed to, but that presented problems related to immediate temporary actions and bookkeeping.

The nice thing about doing the account way is that it could have removed the need to make changes to the bank module. However, problems with doing that include:
* We can't just have it return 0 for unlocked coins because we don't want sanctioned accounts to be able to delegate their funds.
* Instituting immediate temporary sanctions/unsanctions becomes more complex, specially when considering cases where an account has both a temporary immediate sanction and temporary immediate unsanction.

**Comment by SpicyLemon** [comment]
In the future, hopefully the SDK has adopted the injectable bank restictions PR: https://github.com/cosmos/cosmos-sdk/pull/14224

Once that's in the SDK, it should be pretty easy to refactor these changes as such:
1. Create a new restriction function in the sanction keeper that matches the required signature and returns an error if `IsSanctionedAddr` returns false.
2. Inject that restriction into the bank keeper instead of injecting the sanction keeper.
3. Remove the injectable sanction keeper stuff from the bank module.

At that point, we might even consider moving the whole module out of our SDK fork and into provenance (or maybe to its own repo as an importable library).

**Comment by SpicyLemon** [comment]
I still have some unknown scenarios to iron out and check:
1. When an account is sanctioned, should that trigger an undelegation?
2. Are we legally allowed to slash funds in sanctioned accounts?
3. Does anything special need to happen if a validator account is sanctioned?
4. Where are delegated funds kept and do we need to make special considerations for them?

**Comment by SpicyLemon** [comment]
There's some complex stuff going on with the immediate temporary sanctions/unsanctions.

When a `MsgSanction` or `MsgUnsanction` is approved, all temporary entries for each address involved are removed (regardless of gov prop id).

When a prop with a `MsgSanction` and/or `MsgUnsanction` doesn't pass, all temporary entries for just that proposal are removed. If there are other temporary entries for any of the addresses involved that are from other props, those will remain.

Basically, a passed proposal supersedes temp stuff still being voted on.

Furthermore, if an account has multiple temporary entries, the one from the largest prop id takes precedence.

For example, a `MsgSanction` proposal with a huge deposit has account A (say its proposal id is 3). An immediate temporary sanction is created on account A. Then, a `MsgUnsanction` with a huge deposit is submitted containing account A (say its proposal id is 5). An immediate temporary unsanction is created on account A. And since the temp unsanction is from an higher proposal id, it takes precedence; so account A is in an unsanctioned state. After that, prop 3 passes. The two temporary entries are removed and a permanent sanction is placed on A. If later, prop 5 passes, that permanent sanction is removed, but until then, account A is sanctioned. If later, prop 5 fails, nothing happens. Well, technically, it would attempt to remove the temporary unsanction, but since it's already been removed, nothing would happen.

In that example, if, somehow, prop 5 fails/is rejected/terminates (i.e. ends but doesn't pass) while prop 3 is still being voted on, the temporary unsanction from it would be deleted leaving the temporary sanction in place.

Lastly, messages in a proposal are processed in order. So if, for some reason, a proposal has a `MsgSanction` and `MsgUnsanction`, and one or more addresses are listed in both, the last msg an address appears in takes precedence for that address.

There were several decisions I made while designing it this way.
1. I chose to use the gov prop id for temp precedence instead of something else like voting start, voting end, etc. because its a) easier to work with than a time, b) easier to find (e.g. its the only thing passed to the gov hook functions), c) it's handier in that it allows looking up the proposal directly (whereas a time/height does not), d) it can't change during the life of a proposal.
    I recognize that it's not a certainty that gov prop `a` finishes its voting before `a+1`, but I figured that 1) that's a really rare case, 2) creating a solution without this inconsistency would be overly complex and inefficient. So I decided this worry was negligible.
2. When a proposal passes, it clears out all temp entries for any addresses involved. I figured a passing vote should carry more weight than a large deposit.
3. When a proposal doesn't pass, it clears out the temp entries only for that proposal. I did this so that someone couldn't game the system by submitting a follow-up immediate sanction (as a response to a previous, ongoing immediate sanction), then somehow ending/failing that gov prop early. By only removing temp entries associated with a proposal, the first immediate sanction would stay in place.
4. In some cases, a governance hook can be called after a prop has been deleted. For that reason, when cleaning up temp entries (e.g. prop didn't pass), an index recorded in state is used instead of getting the addresses from the message (which might not exist anymore). Creating this index doubles the amount of data needed in state for temporary entries, though. I felt this was acceptable due to the rarity, short life-span, and small size of such entries. A temp record key is 10 bytes + the address, and it's value is 1 byte. A temp index record is the same. For 32 byte addresses, that's 2 * 43 = 86 bytes for the both. For 20 byte addresses, that's 2 * 31 = 62 bytes for both.
    Having the index also allows direct action for a proposal. Without it, all temp entries would need to be iterated when a prop fails in order to find the ones from that prop (that might not actually exist). And since that processing would happen in a `BeginBlock`, I felt it prudent to use the extra space to save the processing. Extra complexity could have been added to only do that processing when the proposal is deleted, but chose to stick with this for the simplicity.
